### ISSUE TITLE
Modules.ps1

### ISSUE THREAD

**Issue by adamdriscoll** [ironmansoftware/powershell-universal#1522]
### Summary of the new feature / enhancement

We should have some sort of Modules.ps1 file that includes a list of modules to install when an instance is starting up. This prevents having to include the modules directly in the repository.

**Comment by rwidmark** [comment]
Don't forget to let it sync with git also, that's critical for many of us. But honestly I don't see the winning of this feature as you can simply just add it in the dashboards.ps1

**Comment by mabster** [comment]
@rstolpe This feature will be really handy for people like me who have dependencies on large binary modules (like Microsoft.Graph) and don't want them syncing into the GitHub repo. If PSU is able to install those modules outside of the repository folder on startup, and include them in the module path, that'd be ideal.

@adamdriscoll It'd be great if PSU also checked for the latest version of each module and updated them on startup before they get imported into any scripts! The startup script I'm using right now does that with some success.

**Comment by Viajaz** [comment]
@mabster Just keep in mind that PowerShellGet v2 is very slow (https://github.com/PowerShell/PowerShellGet/discussions/607) and PowerShellGet v3 has not been released, updating modules before start-up could slow down start-up dramatically. When I run an `Update-Module` on our PSU server, it usually takes over an hour.

**Comment by mabster** [comment]
@Viajaz Wow! Hopefully Adam can find a nice way to do it. My script doesn't use Update-PSResource - it uninstalls the module first if there's a newer version available and then uses Save-PSResource to install it anew. Takes a little under two minutes to do 15 modules.

**Comment by Viajaz** [comment]
@mabster I believe `*-PSResource` are PowerShellGet v3 Cmdlets.

**Comment by mabster** [comment]
@Viajaz Yeah PSU uses PowerShellGet v3 internally. I wasn't able to use Save-Module in my startup script and Adam set me straight as to why.

**Comment by rwidmark** [comment]
> @rstolpe This feature will be really handy for people like me who have dependencies on large binary modules (like Microsoft.Graph) and don't want them syncing into the GitHub repo. If PSU is able to install those modules outside of the repository folder on startup, and include them in the module path, that'd be ideal.
> 
> 
> 
> @adamdriscoll It'd be great if PSU also checked for the latest version of each module and updated them on startup before they get imported into any scripts! The startup script I'm using right now does that with some success.

Now I understand, I thought it was something else.
You might find something in my modile update script useful.

https://github.com/rstolpe/PowerShell-Scripts/blob/main/Windows/Confirm-NeededModules.md

**Comment by adamdriscoll** [comment]
I've implemented a super simple `initialize.ps1` script that should help be a stop-gap for `modules.ps1` while we build something a bit more managed. 

The idea is that in v3.3 PSU starts the webserver immediately. This means we can report startup state. This helps with scenarios like IIS, Docker and Azure WebApps where if we aren't running fast enough and responding to web requests, the host will kill the process and start it up again. The PSU web server will now be available almost immediately. 

This also means we can do longer running processes in start up (if necessary). `initialize.ps1` will run before anything is configured. This will ensure that you can install modules, register vaults, etc before PSU starts doing its thing. 

This is an example of what is displayed in the browser while we are running some initialization logic. 

![image](https://user-images.githubusercontent.com/1256531/187705739-a5c779ec-df13-4625-93af-a2cfbb868fb2.png)

**Comment by rwidmark** [comment]
@adamdriscoll nice, but is it possible to get http 503 or similar in return when it's in that state? Don't want my load balancer to think it's up and running

**Comment by rwidmark** [comment]
> I've implemented a super simple `initialize.ps1` script that should help be a stop-gap for `modules.ps1` while we build something a bit more managed. 
> 
> 
> 
> The idea is that in v3.3 PSU starts the webserver immediately. This means we can report startup state. This helps with scenarios like IIS, Docker and Azure WebApps where if we aren't running fast enough and responding to web requests, the host will kill the process and start it up again. The PSU web server will now be available almost immediately. 
> 
> 
> 
> This also means we can do longer running processes in start up (if necessary). `initialize.ps1` will run before anything is configured. This will ensure that you can install modules, register vaults, etc before PSU starts doing its thing. 
> 
> 
> 
> This is an example of what is displayed in the browser while we are running some initialization logic. 
> 
> 
> 
> ![image](https://user-images.githubusercontent.com/1256531/187705739-a5c779ec-df13-4625-93af-a2cfbb868fb2.png)
> 
> 

So my guess is that it's possible to cache things to psucache with this function? @adamdriscoll

**Comment by rwidmark** [comment]
@Viajaz do you know if it's possible with the int.ps1 file to read in data to PSUcache?

**Comment by mabster** [comment]
Hmm I just learned about "requirements.psd1" with Azure Functions:

https://learn.microsoft.com/en-us/azure/azure-functions/functions-reference-powershell?tabs=portal#dependency-management

It seems like this would be a nice way to structure modules.ps1. Simply return a hashtable of module and version (with version wildcard support) and let PSU do the installation/updating.

I admit this would not be as flexible as my existing solution where I get to decide the target folder etc, but I'd happily trade flexibility for ease of use.

**Comment by adamdriscoll** [comment]
@rstolpe - It should be possible. We aren't actually loading the Universal module directly but it should be on the PSModulePath. You may need to ipmo Universal first and then you can use the cache.

**Comment by adamdriscoll** [comment]
We had another user mention the desire for this and I think we should finally move this forward. Some additional ideas:

- Determine when to install (e.g. git sync, system startup etc)
- Specify if modules should be installed in WinPS and PS7 or both
